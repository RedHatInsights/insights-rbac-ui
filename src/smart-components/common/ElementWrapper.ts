import React, { cloneElement } from 'react';
import { useOutletContext } from 'react-router-dom';

interface ElementWrapperProps {
  path?: string;
  children: React.ReactElement;
}

export const ElementWrapper: React.FC<ElementWrapperProps> = ({ children, path }) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const componentProps: any = useOutletContext();
  return cloneElement(children, { ...componentProps, ...((path && componentProps?.[path]) || {}) });
};

export default ElementWrapper;
