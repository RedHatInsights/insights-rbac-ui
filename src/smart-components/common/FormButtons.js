import React from 'react';
import PropTypes from 'prop-types';
import { ActionGroup, Button } from '@patternfly/react-core';
import { isEmpty } from 'lodash';
import useFormApi from '@data-driven-forms/react-form-renderer/dist/cjs/use-form-api';
import FormSpy from '@data-driven-forms/react-form-renderer/dist/cjs/form-spy';

const FormButtons = ({ dirtyFieldsSinceLastSubmit, submitSucceeded, pristine }) => {
  const { onCancel } = useFormApi();
  const noChanges = isEmpty(dirtyFieldsSinceLastSubmit) || (!submitSucceeded && pristine);
  return (
    <ActionGroup className="pf-u-mt-0">
      <Button type="submit" isDisabled={noChanges} variant="primary">
        Submit
      </Button>
      <Button variant="link" onClick={() => onCancel()}>
        Cancel
      </Button>
    </ActionGroup>
  );
};

FormButtons.propTypes = {
  dirtyFieldsSinceLastSubmit: PropTypes.arrayOf(
    PropTypes.shape({
      [PropTypes.string]: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
    })
  ),
  submitSucceeded: PropTypes.bool,
  pristine: PropTypes.bool,
  onCancel: PropTypes.func,
};

const FormButtonWithSpies = (formProps) => (
  <FormSpy
    subscription={{
      pristine: true,
      submitSucceeded: true,
      dirtyFieldsSinceLastSubmit: true,
    }}
  >
    {(props) => <FormButtons {...props} {...formProps} />}
  </FormSpy>
);

export default FormButtonWithSpies;
