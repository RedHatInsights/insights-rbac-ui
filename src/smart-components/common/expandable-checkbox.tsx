import React from 'react';
import { useFormApi } from '@data-driven-forms/react-form-renderer';
import useFieldApi, { UseFieldApiConfig } from '@data-driven-forms/react-form-renderer/use-field-api';
import { Checkbox, ExpandableSection, FormGroup, Radio, Tooltip } from '@patternfly/react-core';
import OutlinedQuestionCircleIcon from '@patternfly/react-icons/dist/dynamic/icons/outlined-question-circle-icon';

export type ExpandableCheckboxItemType = {
  title: string;
  description: string;
  name: string;
  options?: { title: string; description: string; name: string }[];
};

const ExpandableCheckbox: React.FC<UseFieldApiConfig> = (props) => {
  const formOptions = useFormApi();
  const { items, input } = useFieldApi(props);
  const values = formOptions.getState().values;
  return (
    <FormGroup>
      {items?.map((item: ExpandableCheckboxItemType, key: number) => (
        <Checkbox
          inputClassName="pf-v5-u-mt-xs"
          key={key}
          isChecked={!!values[input.name]?.[item.name]}
          onChange={(_e, value) => {
            const newVal = value && item.options ? item.options?.[0]?.name : value;
            input.onChange({
              ...values[input.name],
              [item.name]: newVal,
            });
          }}
          label={
            <ExpandableSection toggleText={item.title} isIndented>
              {item.description}
              {item.options && (
                <FormGroup role="radiogroup" fieldId={`${item.name}-radiogroup`}>
                  {item.options?.map((option, key) => (
                    <Radio
                      key={key}
                      name={`${item.name}-radio`}
                      isChecked={values[input.name]?.[item.name] === option.name}
                      onChange={() => {
                        input.onChange({
                          ...values[input.name],
                          [item.name]: option.name,
                        });
                      }}
                      label={
                        <React.Fragment>
                          {option.title}{' '}
                          {option.description && (
                            <Tooltip content={option.description}>
                              <OutlinedQuestionCircleIcon />
                            </Tooltip>
                          )}
                        </React.Fragment>
                      }
                      id={`${item.name}-option-${key}`}
                    />
                  ))}
                </FormGroup>
              )}
            </ExpandableSection>
          }
          id={`${item.name}-checkbox`}
          name={item.name}
        />
      ))}
    </FormGroup>
  );
};

export default ExpandableCheckbox;
