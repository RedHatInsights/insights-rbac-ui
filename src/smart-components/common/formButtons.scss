.ins-m__action-group {
  margin-top: 0px !important;
}
