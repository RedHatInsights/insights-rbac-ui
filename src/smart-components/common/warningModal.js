import React from 'react';
import PropTypes from 'prop-types';
import { Modal, ModalVariant, Button } from '@patternfly/react-core';
import { ExclamationTriangleIcon } from '@patternfly/react-icons';
import './warningModal.scss';

export const WarningModal = ({ type, isOpen, onModalCancel, onConfirmCancel }) => (
  <Modal
    title={
      <span className="ins-c-wizard__cancel-warning-header">
        <ExclamationTriangleIcon size="md" className="ins-c-wizard__cancel-warning-header--icon" />
        Exit {type} creation?
      </span>
    }
    variant={ModalVariant.small}
    className="ins-c-wizard__cancel-warning"
    isOpen={isOpen}
    onClose={onModalCancel}
    actions={[
      <Button key="confirm" variant="primary" onClick={onConfirmCancel}>
        Exit
      </Button>,
      <Button key="cancel" variant="link" onClick={onModalCancel}>
        Stay
      </Button>,
    ]}
    isFooterLeftAligned
  >
    <span> All inputs will be discarded.</span>
  </Modal>
);

WarningModal.propTypes = {
  type: PropTypes.string,
  isOpen: PropTypes.bool,
  onModalCancel: PropTypes.func,
  onConfirmCancel: PropTypes.func,
};
