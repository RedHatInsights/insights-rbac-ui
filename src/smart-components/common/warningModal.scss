.ins-c-wizard__cancel-warning-header {
    display: flex;
    align-items: center;
    &--icon {
        fill: var(--pf-global--warning-color--100);
        margin-right: var(--pf-global--spacer--md);
    }
}
