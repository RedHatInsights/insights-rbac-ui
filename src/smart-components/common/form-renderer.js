import React from 'react';
import PropTypes from 'prop-types';
import Pf4SelectWrapper from '../../presentational-components/shared/pf4-select-wrapper';
import FormButtons from './FormButtons';
import FormTemplate from '@data-driven-forms/pf4-component-mapper/dist/cjs/form-template';
import TextField from '@data-driven-forms/pf4-component-mapper/dist/cjs/text-field';
import Textarea from '@data-driven-forms/pf4-component-mapper/dist/cjs/textarea';
import ReactFormRender from '@data-driven-forms/react-form-renderer/dist/cjs/form-renderer';
import componentTypes from '@data-driven-forms/react-form-renderer/dist/cjs/component-types';

const FormRenderer = ({ formTemplateProps, ...props }) => (
  <ReactFormRender
    componentMapper={{
      [componentTypes.TEXT_FIELD]: TextField,
      [componentTypes.TEXTAREA]: Textarea,
      [componentTypes.SELECT]: Pf4SelectWrapper,
    }}
    FormTemplate={(props) => <FormTemplate {...formTemplateProps} {...props} FormButtons={FormButtons} />}
    {...props}
  />
);

FormRenderer.propTypes = {
  formTemplateProps: PropTypes.object,
};

export default FormRenderer;
