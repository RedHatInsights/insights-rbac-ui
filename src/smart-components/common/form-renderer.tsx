import React from 'react';
import FormButtons from './FormButtons';
import FormTemplate from '@data-driven-forms/pf4-component-mapper/form-template';
import TextField from '@data-driven-forms/pf4-component-mapper/text-field';
import Textarea from '@data-driven-forms/pf4-component-mapper/textarea';
import ReactFormRender from '@data-driven-forms/react-form-renderer/form-renderer';
import componentTypes from '@data-driven-forms/react-form-renderer/component-types';
import FormTemplateCommonProps from '@data-driven-forms/common/form-template';

const FormRenderer: React.FC<{ formTemplateProps?: FormTemplateCommonProps } & FormTemplateCommonProps> = ({ formTemplateProps, ...props }) => (
  <ReactFormRender
    componentMapper={{
      [componentTypes.TEXT_FIELD]: TextField,
      [componentTypes.TEXTAREA]: Textarea,
    }}
    FormTemplate={(props) => <FormTemplate {...formTemplateProps} {...props} FormButtons={FormButtons} />}
    {...props}
  />
);

export default FormRenderer;
