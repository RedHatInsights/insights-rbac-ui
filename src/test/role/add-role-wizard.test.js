import React from 'react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { shallowToJson } from 'enzyme-to-json';

import { MemoryRouter, Route } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';
import AddRoleWizard from '../../smart-components/role/add-role/add-role-wizard';
import { ADD_NOTIFICATION } from '@redhat-cloud-services/frontend-components-notifications/index';
import { mount } from 'enzyme/build/index';
import { WarningModal } from '../../smart-components/common/warningModal';

describe('<AddRoleWizard />', () => {
  let initialProps;
  let initialState;
  const middlewares = [thunk, promiseMiddleware, notificationsMiddleware()];
  let mockStore;

  const RoleWrapper = ({ store, children }) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={['/roles/add-role/']}>{children}</MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {};
    initialState = {
      roleReducer: {
        roles: [
          {
            label: 'foo',
            value: 'bar',
          },
        ],
      },
    };
    mockStore = configureStore(middlewares);
  });

  it('should render correctly', () => {
    const store = mockStore(initialState);
    const wrapper = shallow(
      <RoleWrapper store={store}>
        <AddRoleWizard {...initialProps} />
      </RoleWrapper>
    ).dive();

    setImmediate(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it('should show a warning modal on Cancel', () => {
    const store = mockStore(initialState);

    const wrapper = mount(
      <RoleWrapper store={store}>
        <Route path="/roles/add-role/" render={() => <AddRoleWizard {...initialProps} />} />
      </RoleWrapper>
    );
    expect(wrapper.find(WarningModal).getElement().props.isOpen).not.toBeTruthy();
    const input = wrapper.find('input#name');
    input.getDOMNode().value = 'foo';
    input.simulate('change');
    wrapper.update();
    wrapper.find('.pf-m-link').simulate('click');
    wrapper.update();
    expect(wrapper.find(WarningModal).getElement().props.isOpen).toBeTruthy();
  });

  it('should not show a warning modal on Cancel when clean', (done) => {
    const store = mockStore(initialState);

    const wrapper = mount(
      <RoleWrapper store={store}>
        <Route path="/roles/add-role/" render={() => <AddRoleWizard {...initialProps} />} />
      </RoleWrapper>
    );
    const expectedActions = expect.arrayContaining([
      expect.objectContaining({
        type: ADD_NOTIFICATION,
        payload: expect.objectContaining({ title: 'Creating role was canceled by the user', variant: 'warning' }),
      }),
    ]);
    expect(wrapper.find(WarningModal).getElement().props.isOpen).not.toBeTruthy();
    wrapper.find('.pf-c-button.pf-m-primary').simulate('click');
    wrapper.update();
    wrapper.find('.pf-m-link').simulate('click');

    setImmediate(() => {
      expect(store.getActions()).toEqual(expectedActions);
      expect(wrapper.find(WarningModal)).toHaveLength(0);
      done();
    });
  });
});
