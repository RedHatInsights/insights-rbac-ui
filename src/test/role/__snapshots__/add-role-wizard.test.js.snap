// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddRoleWizard /> should not show a warning modal on Cancel when clean 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        "/roles/add-role/",
      ]
    }
  >
    <withRouter(Connect(AddRoleWizard)) />
  </MemoryRouter>
</ContextProvider>
`;
